### PR TITLE
Restored `-remoteHost` support and fix for nodes behind reverse proxy

### PR DESCRIPTION
**Pull request by kinjelom**
1. Restored `-remoteHost` CLI option support, lost in revision: 2fb93404832400c5c02fdac6593b19d19e582d8d (@barancev "Composition instead of inheritance in CLI options")
2. Support for nodes behind reverse proxy (Cloud Foundry GoRouter in my case) https://github.com/SeleniumHQ/selenium/issues/5373

- [X] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#step-6-sign-the-cla)

---
This change is [Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/6627)

### CODE REVIEW

File: java/server/src/org/openqa/grid/common/RegistrationRequest.java [hunk 1]
```diff
@@ -203,6 +203,9 @@ public static RegistrationRequest build(GridNodeConfiguration configuration, Str
     if (pendingConfiguration.port != null) {
       pendingRequest.configuration.port = pendingConfiguration.port;
     }
+    if (pendingConfiguration.remoteHost != null) {
+      pendingRequest.configuration.remoteHost = pendingConfiguration.remoteHost;
+    }
```
Comment (diemol) [inline java/server/src/org/openqa/grid/common/RegistrationRequest.java]: This is not really needed, the line added in `GridNodeConfiguration.java` is what really fixes the `remoteHost` regression.

**Comment by diemol** [comment]
Hi @wildloop, thanks for this PR.

I added a small comment about the `remoteHost` regression.

About the reverse proxy changes, I am not sure if we should add this without tests. Could you please add some tests to cover this? Moreover, this would help in your use case, are we sure that this does not break any other use cases?
What do you think about the reverse proxy change @barancev @shs96c?

**Comment by wildloop** [comment]
@diemol what about enabling this feature with special CLI option `-supportReverseProxy true` or maybe if `-remoteHost` is a domain name?

**Comment by diemol** [comment]
@wildloop we don't want to add new flags or features but only the bug fix, the reason for this is that we expect the next release is the last one of the current Grid. We are currently working out in a new Grid, which welcomes suggestions like the one you would like to add. 
If you would like to check the in-progress development code of the new Grid, check [here](https://github.com/SeleniumHQ/selenium/tree/master/java/server/src/org/openqa/selenium/grid). Also, you can join us in the [IRC/Slack channel](https://seleniumhq.herokuapp.com) where we can discuss new things as well.

Therefore, could you please leave only the bug fix in this PR so we can merge it? Thanks!

**Comment by wildloop** [comment]
@diemol Support for reverse proxy is basically a fix too. How do you use `remoteHost`? I think that it works only when `remoteHost` is IP address or the same domain as hub's domain. In these cases change of a `Host` HTTP header doesn't matter - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host
That fix gives possibility to run the grid's nodes on a remote host indicated by domain address.

File: java/server/src/org/openqa/grid/common/RegistrationRequest.java (hunk 1 above)
Comment (wildloop) [reply java/server/src/org/openqa/grid/common/RegistrationRequest.java]: You're right :) corrected: https://github.com/SeleniumHQ/selenium/pull/6627/commits/91e71499f37dc1852db0df9bbbb7480acbf0cef2

**Comment by wildloop** [comment]
Ready to manual tests:
- https://github.com/wildloop/selenium/releases/download/3.141.59-SNAPSHOT/selenium-server-standalone-3.141.592-SNAPSHOT.jar (piver has even more power than semver.org ;) )
- https://hub.docker.com/r/wildloop/selenium-hub/
- https://hub.docker.com/r/wildloop/selenium-node-chrome/
- https://hub.docker.com/r/wildloop/selenium-node-firefox/

**Comment by diemol** [comment]
@wildloop just to circle back on the previous point and make it more clear, we want to avoid making changes to support new use cases in the Grid, especially to centric classes like `TestSession.java`. 

We understand that this might be a bug fix for your use case, but we are not aware of what side effects this could bring to other users. We prefer to move things like this to the new Grid, where we welcome you to participate in its development by making suggestions and PRs.

In addition, for future reference, it is more desirable to add tests to changes like this one.

So please, leave only the bug fix in the PR so we can merge it and give you credit for it. Thanks!

**Comment by wildloop** [comment]
@diemol done: https://github.com/SeleniumHQ/selenium/pull/6627/commits/6cfb57662d62d230d559f236cbc2cc1a090aa2ef

**Comment by diemol** [comment]
Merged as https://github.com/SeleniumHQ/selenium/commit/07a116aa32e6b464567a331ed9d1c4ffbee104bb, thanks for your contribution @wildloop!

**Comment by wildloop** [comment]
Thanks! 

-----
If someone needs support for reverse proxy before the new grid will be released, he can use my forks:
- [`selenium-server-standalone` source code](https://github.com/wildloop/selenium/tree/reverse-proxy-support)
- [`selenium-docker` source code](https://github.com/wildloop/selenium/tree/reverse-proxy-support)

Binaries:
- [selenium-server-standalone-3.141.592-SNAPSHOT.jar](https://github.com/wildloop/selenium/releases/download/3.141.59-SNAPSHOT/selenium-server-standalone-3.141.592-SNAPSHOT.jar)
- [`docker pull wildloop/selenium-hub`](https://hub.docker.com/r/wildloop/selenium-hub/)
- [`docker pull wildloop/selenium-node-chrome`](https://hub.docker.com/r/wildloop/selenium-node-chrome/)
- [`docker pull wildloop/selenium-node-firefox`](https://hub.docker.com/r/wildloop/selenium-node-firefox/)

**Comment by lionlamb** [comment]
Hi @wildloop ,
I have been using your binaries and manifest.yaml examples and pushed to CF, however the node still not able to connect with hub and talk to https. I am wondering which part had I done wrong. Do you mind to shade some light?

**Comment by wildloop** [comment]
@lionlamb Please try to connect to http://selenium-node-chrome-1.mycloudabc.cloud.com - HTTP protocol is disabled by default but selenium client doesn't support HTTPS (as I remember).
Next you can set `GRID_DEBUG: true` and share some logs.

You can try to configuration with [CF Internal Routing / Domains](https://docs.cloudfoundry.org/devguide/deploy-apps/routes-domains.html#internal-routes) and [C2C Networning](https://docs.cloudfoundry.org/concepts/understand-cf-networking.html#policies) that should simplify configuration and may works without hacking ;) ... but I didn't  try one.

**Comment by lionlamb** [comment]
@wildloop 

Thanks for your kind reply.  I have tried many hours and checked the logs many times but still no luck. 

First, I setup both hub and node respectively again from the changes you share, it shows the following error:

`2019-05-12T22:13:30.51+0800 [xxxxxx] OUT 14:13:30.512 INFO [SelfRegisteringRemote$1.run] - Couldn't register this node: The hub is down or not responding: Failed to connect to selenium-hub.apps.internal/10.177.166.161:8080`

And i checked the hub, it seems like created at port 4444 automatically, so I update the manifest to try up. still getting the above error (just the 8080 replaced with 4444), here is the log info from hub

` 2019-05-12T20:18:26.81+0800 [xxxxxx] OUT 12:18:26.815 INFO [Hub.start] - Clients should connect to http://10.177.166.161:4444/wd/hub`

Then, I tried to do internal routing as per mentioned by you :) and got the message that the hub and node already exists:

```
cf map-route selenium-hub apps.internal --hostname selenium-hub
Creating route selenium-hub.apps.internal for mycloudabc / space seleniumTest as cloudtest@mycloudabc.com...
OK
Route selenium-hub.apps.internal already exists
```

I am curious do we need to EXPOSE the port 4444 in the docker image?

**Comment by wildloop** [comment]
What about cf network policies?
Did you create policy to allow direct network traffic between nodes and hub?
https://cli.cloudfoundry.org/en-US/cf/add-network-policy.html

**Comment by lionlamb** [comment]
@wildloop I didn't because SAP Cloud Foundry doesn't support it

**Comment by wildloop** [comment]
So, go back to prev config and enable debug logs.
You can also connect to container using `cf ssh app-name` and test connection to hub/nodes (e.g. using netcat `nc -tvn <ip> <port>`).

**Comment by lionlamb** [comment]
Yup, enabled the debug logs & Using netcat from node to test connection 

```
seluser@54dfef6b-bf7d-6c54-76cf-5hdg:~$ nc -tvn 10.177.166.161 4444
(UNKNOWN) [10.177.166.161] 4444 (http-alt) : Connection refused
```

**Pull request closed without merging**